TEP 0109: Better structured provenance retrieval in Tekton Chains

ywluogg opened the pull request:
Start the discussions and reviews regarding to support structured results and params in Tekton Chains

ywluogg commented:
/assign @wlynch @afrittoli

ywluogg commented:
This is a draft that I have. I'm currently working on changes to add PipelineRun into scope as well.

wlynch commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
````diff
@@ -0,0 +1,478 @@
+    digest:
+      type: string
+```
+The required formats for src_type is one of `git`, `hg`, `svc`, `baz`, `cvs`, `perforce`. The source types can be easily extended when more VCS are supported. The required formats for uri should follow [ResourceURI](https://github.com/in-toto/attestation/blob/main/spec/field_types.md#field-type-definitions). The required format for alg is one of pre-defined algorithm names in intoto attestation [ ...
````
Is there a reason why we need to constrain the source type? 👀

ywluogg replied:
Ah maybe my wording is hinting this, but our source types are not constrained. This can be extended later. I changed my wording to make it clearer

wlynch commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,478 @@
+
+#### Version Control Sources
+
+Based on the use cases and data storage within provenance described above, it makes sense to have the source object to be the following because all version control sources will be grinded down into the requirements in the Predicate’s materials:
```
Just to be explicit here, this isn't a new type we're inventing here right? What we're really trying to do is standardize on SLSA's `materials` field and defintion?

ywluogg replied:
Yes. They are not invented from here - these structures are similar because they follow the standards of `materials` in SLSA

wlynch replied:
Since theses are all just variants of the same message schema, perhaps it would make more sense to just define 1 result type and list the various usages (e.g. git, oci, python, ...) under an examples section?

wlynch commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
````diff
@@ -0,0 +1,478 @@
+    alg:
+      type: string
+```
+The `name` is the OCI image name, for example, `gcr.io/test/foo`. The `digest` will be the image digest in the form of `69e70a79f2d41ab5d637de98c1e0b055206ba40a8145e7bddb55ccc04e13cf8f`. The `alg` follows the same requirements for VCS `alg` field.
````
Same as above, this is really just standardizing on SLSA's `subject` field, right?

ywluogg replied:
Yep!

wlynch commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,478 @@
+
+#### Storage of Signable Target Provenance Schema
+
+Each target provenance schema will be stored in a file as a serialized json object. They will be loaded and mounted into the Chains controller image at volumeMountPath, for example, `tekton/target_schemas`. A singleton variable will load them into structs for later use. The naming will be the type hinting marker. For example, VCS schema file will be called `chains-vcs` and it should be stored in  ...
```
This needs more detail - volume sprawl is already a concern in Tekton, and I'd be hesitant to add more. There were interesting ideas in https://github.com/tektoncd/community/pull/521 that may be of interest here (I was particularly a fan of the OIDC based communication to tackle some of these problems).

ywluogg replied:
I changed it to be included in the controller image. They will be predefined json schemas living in source code.

wlynch commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,478 @@
+
+#### Storage of Signable Target Provenance Schema
+
+Each target provenance schema will be stored in a file as a serialized json object. They will be loaded and mounted into the Chains controller image at volumeMountPath, for example, `tekton/target_schemas`. A singleton variable will load them into structs for later use. The naming will be the type hinting marker. For example, VCS schema file will be called `chains-vcs` and it should be stored in  ...
```
> The naming will be the type hinting 

I'd recommend against this, since if we're using results these will be user controlled values.

Possibly a crazy idea (not even 100% sold on this myself), but perhaps provenance information shouldn't belong in results and instead should be their own independent field? 🤔

ywluogg replied:
Updated the proposal section to reflect what we discussed

ywluogg commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,478 @@
+
+## Proposal
+
+Currently we are using type-hinting to retrieve needed provenance info from TaskRun’s params and results. The proposal aims to leverage the new features of structured params and results described in [TEP 0075](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md) and [TEP 0076](https://github.com/tektoncd/community/blob/main/teps/0076-array-result-types.md) a ...
```
Can you explain a bit more about ideas of such tool? Do you mean we provide an image for users or a piece of module in Chains?

wlynch replied:
In my mind tool == any task (i.e. a git task). The Tasks would need to be responsible for outputting the relevant provenance for their given artifact type.

Chains wouldn't need to be aware of any particular tasks/images - it would just consume the provenance output for each task (similar to the result based mechanism it uses today).

ywluogg replied:
To reflect what we discussed offline, I updated the doc in proposal section. I added flexibility that the provenance can come from results and other new fields that's only for trusted tasks

lcarva replied:
I like the idea of having something that generates the structure response because it allows users to extend support for a particular VCS without having to make any changes to Chains. It's just another Task that needs to be written.

With [TEP-84](https://github.com/tektoncd/community/blob/main/teps/0084-endtoend-provenance-collection.md) in place, post-build policies can be put in place to ensure that the Task that produced the VCS information is trusted.

ywluogg commented:
/assign @dibyom

pritidesai commented:
/kind tep

lcarva commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,544 @@
+            type: string
+          digest:
+            type: string
+          alg:
```
It's not clear that this is the algorithm of the digest. Maybe `digest_ago` is more specific? Alternatively `digest` could be a specially formatted string that includes both information, e.g. `<algo>:<digest>`.

ywluogg replied:
I changed the name to `digest_ago`. I'm still keeping them separate to give more flexibilities.

dibyom commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,544 @@
+
+#### Parsing Results and Params from TaskRuns
+
+The defined results and params are in object types, and will be retrieved from TaskRun params and results. From the syntax of how the results / params look, the type of those, whether they are strings, objects or arrays can be determined. We need to validate the targets’ structures against defined result structures. Taking VCS as an example, when a `CHAINS-VCS` is found, the target needs to be an ...
```
Based on the discussion above, is the plan to still rely on user params or rather have the task output a result that expands the user input into a structured result that Chains can use?

ywluogg replied:
Thanks for reminding this! Now I removed all the proposals related to params.

ywluogg commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,544 @@
+    kind: Task
+    spec:
+      params:
+        - name: CHAINS-VCS
```
SG! I'm changing the current way to have the naming to be not scoped only Chains. Another option is to we require users to provide a result in array type, with name consistent with the signable target, and have a list of result names that are under that category. An example can be like:
```
results:
        - name: SIGNABLE-VCS
          value: ["FOO", "BAR"]
        - name: FOO
           value:
             name: gcr.io/somerepo/someimage
             digest: abc
             alg: sha256
```
And the result that contains the provenance still need to be aligned with the schema we defined in Chains

ywluogg commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,544 @@
+        - description: artifact image
+      ...
+      results:
+        - name: IMAGE_DIGEST
```
Added!

ywluogg commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,544 @@
+            type: string
+      ...
+      results:
+        - name: CHAINS-OCI_IMAGE
```
Ah yes I changed the naming again. I also wanted to discuss another proposal to let users don't have to name the results in a certain way

wlynch commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,565 @@
+    spec:
+      ...
+      results:
+        - name: SIGNABLE-VCS_$(NAME)
```
is `$(NAME)` actual interpolation, or just a placeholder?

ywluogg replied:
Ah this is just a placeholder in the doc.

wlynch commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
````diff
@@ -0,0 +1,565 @@
+It will be more flexible if users can define the provenance following the defined schemas, but don't necessarily have to follow the naming patterns. Users can provide a result in array type, with name consistent with the signable target, and have a list of result names that are under that category. An example can be like:
+``` yaml
+results:
+        - name: SIGNABLE-VCS
+          value: ["FOO", "BAR"]
````
2 alternatives that are probably worth exploring:

1. Consolidate to 2 results - `INPUTS` / `OUTPUTS` (similar to how `IMAGES` is used today)
2. Don't rely on results at all and instead introduce similar fields into the TaskRun status.

I think my preference long-term would be 2 - though we can also hash out what intermediate steps we would need to get there.

ywluogg replied:
1. Are you trying to suggest something like this?
```
results:
- name: SIGNABLE_INPUTS
          type: array
```
```
results:
- name: SIGNABLE_OUTPUTS
          type: array
```

Then we have this in TaskRun:
```
results:
- name: SIGNABLE_INPUTS
   value:
   - RESULT_1
   - RESULT_2

- name: SIGNABLE_OUTPUTS
   value:
   - RESULT_3
   - RESULT_4

```
But I think it still matters if we can know the type of the signable target (mave or python), and that would involve something like this with nested objects:
- name: SIGNABLE_INPUTS
   value:
    - name: RESULT_2
       signable_type: VCS
    - name: RESULT_3
       signable_type: IMAGE

Considering nested objects are out of scope for TEP 75, I would probably consider as future support options, but we can also embed this nested object support into option 2.

The current schema has a `signable_type` field that indicates whether the result is an input or an output.

2. I agree that this should be either ultimately put under a TaskRun field or getting this assured that the results are generated by some trusted tasks

ywluogg replied:
Discussed offline to refactor the schema. See doc for details

wlynch commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
````diff
@@ -0,0 +1,565 @@
+```
+Tekton Chains will follow [TEP 0075](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md) and [TEP 0076](https://github.com/tektoncd/community/blob/main/teps/0076-array-result-types.md) and support the sets of json schemas and described in the documents. In this proposal, the expected formats of results are also defined.
+
+As mentioned above, the signable targets can be split into inputs (sources) and outputs (artifacts). It's important that Chains are able to determine which signable targets are inputs, and which are outputs. We can enforce all the schemas need to have a string field `prov_type`, which can be either `input` or `output`, which indicates Chains to either put them into `subjects` or `materials`.
````
I'm having trouble following how we're distinguishing inputs and outputs at the moment.

e.g. how would we distinguish "I used this image" vs "I pushed this image"?

ywluogg replied:
I'm trying to say that the result should have `prov_type` to indicate this. Are you trying to address another thing?

ywluogg replied:
Discussed offline. We are going to have SIGNABLE_INPUTS and SIGNABLE_OUTPUTs in results

wlynch commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,565 @@
+
+#### Storage of Signable Target Provenance Schema
+
+Each target provenance schema will be stored in a file as a serialized json object. They will be included into the Chains controller image. A singleton variable will load them for later use. The naming will be the type hinting marker. For example, VCS schema file will be called `vcs` and it should be stored in the file in such format:
```
This is glossing over a lot 😅 

How would provenance be collected from Runs and be provided to Chains? Is https://github.com/tektoncd/community/blob/main/teps/0086-changing-the-way-result-parameters-are-stored.md a prerequisite here, or are you thinking of a different mechanism?

ywluogg replied:
I'm removing this section since we are only having one schema, so storage of those schemas may not be needed for now

ywluogg commented:
/reopen

ywluogg commented:
/assign @wlynch

ywluogg commented:
/assign @afrittoli

pritidesai commented:
/assign @vdemeester

wlynch commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,502 @@
+their own choices?
+-->
+
+### User Experience
```
Would you be able to try out and link to an example of a TaskRun that generates this kind of result output?
This will probably be a good exercise to see what's easy / hard to do as a TaskRun author. 

For example - I don't know if we have a good way of passing in result names to TaskRuns, so generating the INPUTS/OUTPUTS lists might actually be tricky to implement as an author.

ywluogg replied:
Added an example at https://github.com/ywluogg/oci-image-structured-results

wlynch commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,502 @@
+
+## Design Details
+
+### Signable Targets
```
What's the difference between a signable target and an output artifact?

ywluogg replied:
I'm changing all wording to "signable artifacts"

wlynch commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,502 @@
+          description: |
+            The source distribution
+            * uri: resource uri of the artifact.
+            * digest: revision digest in form of algorithm:digest.
```
How are we planning to handle DigestSets?

ywluogg replied:
Can you provide more context here?

wlynch replied:
Right now we have uri as type `string`, but SLSA defines this as `map[string]string`, so this isn't compatible with the intoto spec.

ywluogg replied:
Ah I see your point now. If we want DigestSets to be supported in an organizational way, we need Pipeline Results to have nested objects to be supported. I added a DigestSets section in the doc

wlynch replied:
Do we have a way of representing variable length outputs? e.g. how the `IMAGES` field works today.

I'm trying to think through how things where we may not know the exact # of artifacts produced like multiarch images can be supported. 🤔

ywluogg replied:
That's a super good point! I was investigating in-toto provenance, I think `subjects` and `materials` seem not compatible with multiarch images. `DigestSets` is a `map[string]string`, but the examples in the in-toto repo only shows examples that key is an algorithm and value is the digest itself. I filed https://github.com/in-toto/attestation/issues/105 for tracking that

ywluogg replied:
SLSA team suggested to list them separately and add unique names for those or just add the digest for the image manifest.

wlynch reviewed (changes requested):
minor, but might be worth running this through a formatter like https://prettier.io/ to help clean up some of the line wrapping / spacing.

chuangw6 commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
````diff
@@ -0,0 +1,502 @@
+```
+Tekton Chains will follow [TEP 0075](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md) and [TEP 0076](https://github.com/tektoncd/community/blob/main/teps/0076-array-result-types.md) and support the sets of json schemas and described in the documents. In this proposal, the expected formats of results are also defined.
+
+As mentioned above, the signable targets can be split into inputs (sources) and outputs (artifacts). It's important that Chains are able to determine which signable targets are inputs, and which are outputs. The signable target result names are all captured in either of the results, `ARTIFACT_INPUTS` or `ARTIFACT_OUTPUTS`, which will be two arrays storing the result names that should be captured  ...
````
This line says 
> `ARTIFACT_INPUTS` will be captured in `subjects` and `ARTIFACT_OUTPUTS` will be captured in `materials`.

But the example above is inconsistent. In the example, items from `ARTIFACT_INPUTS` are stored into `materials`, whereas items from `ARTIFACT_OUTPUTS` are stored into `subjects`. Which one is intended?

Also can you please help me understand a bit more about the differences between `ARTIFACT_INPUTS` and `ARTIFACT_OUTPUTS` given that they are all defined in `results`? Thanks

ywluogg replied:
Good catch!

ywluogg replied:
In terms of differences of `ARTIFACT_INPUTS` and `ARTIFACT_OUTPUTS`, they mean that whether an artifact A is used as source or materials in a pipeline in order to generate another artifact B. In this case, A is `ARTIFACT_INPUTS` and B is `ARTIFACT_OUTPUTS`

ywluogg commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,502 @@
+Why should this TEP _not_ be implemented?
+-->
+
+## Alternatives
```
Good point! I added a section for Run `status` in the `Alternatives`.

ywluogg commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,502 @@
+            The source distribution
+            * uri: resource uri of the artifact.
+            * digest: revision digest in form of algorithm:digest.
+            * artifact_type: artifact type categorizes the artifacts.
```
I added more explanations to `artifact_type` in the article. Currently I'm more using this for Chains to monitor metrics. I will mention this to SLSA folks. I also created an issue in Intoto provenance repo: https://github.com/in-toto/attestation/issues/103

ywluogg replied:
Updates from https://github.com/in-toto/attestation/issues/103 - pipeline authors can follow PACKAGE URL format to provide the URI, and this way it can identify the artifact type. However, consider the situation that pipeline authors can't follow the PACKAGE URL format and still want to identify the artifact type, and this `artifact_type ` field will become useful. But I would remove this field, since there is no essential reason to add the field here, and people can still provide URI in PACKAGE URL to identify the artifact type.

ywluogg commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,502 @@
+          description: git commit url
+        - name: FOO_IMAGE
+          type: string
+        - description: artifact image
```
Good catch!

ywluogg commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,502 @@
+            * uri: resource uri of the artifact.
+            * digest: revision digest in form of algorithm:digest.
+            * artifact_type: artifact type categorizes the artifacts.
+          properties:
+            uri:
+              type: string
+            digest:
+              type: string
```
Good catch!

ywluogg commented:
> minor, but might be worth running this through a formatter like https://prettier.io/ to help clean up some of the line wrapping / spacing.

Done!

afrittoli reviewed (commented):
Thank you for this!

Even with TEP-0075 and TEP-0076 implemented, it seems to me that we still lack the ability to express that a result is signable on Tekton Pipeline side. 

I don't like very much the idea of asking Task authors to write the name of the result to a result in a script. The alternative would be to name results with a known prefix, like 'ARTIFACT_INPUT_something_something`. That wouldn't look very nice but it would save the extra result writing in the step.

ywluogg commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,538 @@
+
+The implementation plan has two main components:
+
+1. upgrading Tekton Chains's Tekton Pipeline's version to v0.38, which TEP 75 and 76 will be supported behind alpha feature flag.
```
> Thank you for this!
> 
> Even with TEP-0075 and TEP-0076 implemented, it seems to me that we still lack the ability to express that a result is signable on Tekton Pipeline side.
> 
> I don't like very much the idea of asking Task authors to write the name of the result to a result in a script. The alternative would be to name results with a known prefix, like 'ARTIFACT_INPUT_something_something`. That wouldn't look very nice but it would save the extra result writing in the step.

Thanks for the comments! I think the TEP design was originally based on 'ARTIFACT_INPUT_something_something`, but this reduces the flexibility that Task authors has to follow the naming in a particular format. And that approach can already be applied through a merged implementation https://github.com/tektoncd/chains/pull/465/files. I think the TEP is trying to group multiple infos in a structured way, so I think having two extra results wouldn't hurt that much considering the flexibility in naming it provides.

ywluogg replied:
Thats right!

ywluogg commented:
> Thank you for this!
> 
> Even with TEP-0075 and TEP-0076 implemented, it seems to me that we still lack the ability to express that a result is signable on Tekton Pipeline side.
> 
> I don't like very much the idea of asking Task authors to write the name of the result to a result in a script. The alternative would be to name results with a known prefix, like 'ARTIFACT_INPUT_something_something`. That wouldn't look very nice but it would save the extra result writing in the step.

Thanks for the comments! I think the TEP design was originally based on 'ARTIFACT_INPUT_something_something`, but this reduces the flexibility that Task authors has to follow the naming in a particular format. And that approach can already be applied through a merged implementation https://github.com/tektoncd/chains/pull/465/files. I think the TEP is trying to group multiple infos in a structured way, so I think having two extra results wouldn't hurt that much considering the flexibility in naming it provides. Can you provide more context why it wouldn't look nice if Task authors are setting the naming for the results?

ywluogg commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
````diff
@@ -0,0 +1,538 @@
+ }
+```
+
+The other specific requirements of these fields can be found in [in-toto attestation spec](https://github.com/in-toto/attestation/tree/main/spec). Another thing to be aware is that For `digest`, it has a key for the digest type and has a value for that key. Git, Mercurial, Bazaar and Subversion have full VCS location support. CVS, Perforce are providing this VCS location as plain URLs. For VCS, t ...
````
Ah that was a paragraph addressing some issues in the old design for the schema. I' deleting this paragraph. Thanks for pointing this out!

ywluogg commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,538 @@
+
+It's also beneficial that the schemas defined in Chains can follow the field requirements in in-toto attestations. Currently, the type hinting is following a readable formats that users will find it more readable, such as results being populated in [upload-pypi](https://github.com/tektoncd/catalog/blob/f4708d478ee8fac6b5b68347cde087cb7c1d1b1c/task/upload-pypi/0.1/upload-pypi.yaml) and [jib-maven] ...
+
+Generating the objects that satisfy these schemas, from human readable provenance metadata can be done by writing provenence data using Steps in Tasks into these schema structures in TaskRuns and populate them PipelineRuns' Results, and there is also a discussion around adding a new field `provenance` in PipelineRun and TaskRun, which only allows [Trusted Resources](https://github.com/tektoncd/co ...
```
Done!

ywluogg commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,538 @@
+All below examples will be generating in-toto attestation.
+ 1. Use Git commits as sources, and generates in-toto provenance for a TaskRun that builds an image.
+   
+    A Task example `without` structured results and params support can be like:
```
Done!

ywluogg commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,538 @@
+
+## Motivation
+
+With SLSA being established, there is a rise in demand for achieving richer provenance within attestations. [In-toto](https://github.com/in-toto/attestation) provides a way for supporting richer provenance inside Predicates. Tekton Chains currently supports multiple kinds of attestations and the in-toto attestation formnat is one of the popular ones in the open source community. Tekton Chains cur ...
```
Done!

ywluogg commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
````diff
@@ -0,0 +1,538 @@
+```
+
+And the result that contains the provenance still need to be aligned with the schema we defined in Chains. This way, users don't need to name the results.
+This is not chosen because Chains would know the differences between different types of artifacts.
````
Done

ywluogg commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,538 @@
+
+1. upgrading Tekton Chains's Tekton Pipeline's version to v0.38, which TEP 75 and 76 will be supported behind alpha feature flag.
+
+2. Iterate TaskRun / PipelineRun results for scanning the targeted results.
```
I just reworded the No.2 bullet point. Let me know if that's clearer. Thanks!

ywluogg commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
````diff
@@ -0,0 +1,538 @@
+
+```yaml
+results:
+  - name: ARTIFACT_INPUTS
+    value: ["ARTIFACT-INPUT-NAME-1", "ARTIFACT-INPUT-NAME-1"]
+  - name: ARTIFACT_OUTPUTS
+    value: ["ARTIFACT-OUTPUT-NAME-1"]
````
There are multiple threads such as  #alternatives section in the TEP and https://github.com/tektoncd/community/blob/main/teps/0082-workspace-hinting.md about using a separate field for providing provenance metadata together with Trusted Resources, and the down sides for using results as future field for such purpose is not ideal. So I'm concerned that if I propose to add "signable" field in results, it may not be a future solution that we will be sticking to.

ywluogg replied:
Another thing along this route could be - adding an extra key in the object, such as "artifact_type"? If the results are having this key, and the value is either "input" or "output", then this is a result that Chains is looking for. This doesn't need us to add a field in results in Pipeline.

wlynch replied:
I prefer this over requiring task authors to plumb through result names, though a few caveats:

Putting data in object keys means that we're breaking from the intoto spec - e.g. what happens if a user just outputs a vanilla intoto material? Is intoto compatibility a goal for this proposal (it feels like this is what we're aiming for, but it's not explicitly called out as a goal)?

By putting the field in the result spec, this has the nice property of not having to need to modify the payload the Task needs to output (e.g. it can be a vanilla intoto spec), but if we'll need a spec change it might make more sense to just move forward with the status-based provenance.

ywluogg replied:
I think having the schema being compatible with in-toto provenance should be one of the goals. I also agree that adding a field in results, which requires spec change, may not satisfy our future needs. That said, maybe keeping the results to follow naming type hinting may still be the temporary pattern we should follow before the status-based provenance being landed. WDYT @wlynch @afrittoli ?

wlynch replied:
I can live with it, though I think we should change the arrays - I'm not sure if there's an easy way to have the Task provide these without hardcoding them. I don't think there's a way to derive the values from the [replacement context](https://tekton.dev/docs/pipelines/variables/#variables-available-in-a-task).

ywluogg replied:
> 

Do you mean for Task authors how they can easily add those suffices?

wlynch commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,537 @@
+
+### Using Run Status to generate Provenance Metadata
+
+Results is not an ideal place to have the provenance metadata populated for the artifacts, since Results are provided by Pipeline authors. When people are looking for provenance, the metadata being provided should ideally be trustful and unchangable after generation. How can we garantee the provenance metadata can be trusted? Assuming the Tekton Pipeline being installed and where the Runs are ope ...
```
I think the thing I like the most about not tying provenance to to results is that authors don't need to pre-declare provenance result data - a Task can simply output it without needing it to be explicitly declared in a param. This makes it easier to work with when there's a variable amount of items that will be generated, though at the cost of making it hard to be consumed by other steps (though it's unclear to me whether there's a strong requirement for this at the moment).

ywluogg replied:
I think if users want another step to verify if the produced result is having the correct digest or verify if the uploaded digest matches the computed digest, this becomes critical that the provenance can be consumed in another step

ywluogg replied:
I still find it useful to have provenance metadata retrieval being supported in results, since in status field that would require Trusted Resources, but I see circumstances that users may not be able to come all the way to support Trusted Resources. Having this supported in results can give users more flexibility to whether to choose Trusted Resources or not

wlynch replied:
My thinking is that Tasks could still produce output data into results - it just wouldn't be a requirement and authors can format that data however they want. The benefit is splitting the provenance data that has a predefined API that is meant to be used by other tools from Task results that are meant to be used by other human authors.

We can still expose the raw data into Tasks later on if we wanted to, but we also don't need to commit to that right now.

ywluogg replied:
I see! That's a good point. But how would we allow users to use Chains, if the string format type hinting is something we are looking to deprecate? I think users may still look for a way to use Chains in a more structured way without using Trusted Resources

wlynch replied:
> But how would we allow users to use Chains, if the string format type hinting is something we are looking to deprecate?

How confident are we in the status based provenance as the long term solution? It might make more sense to just move forward with that.
We could support a limited subset short term if needed (i.e. single, predefined artifacts) with the understanding that it's not a general purpose solution and can probably only be used in limited cases.
Since it sounds like we need Pipeline changes anyway to have a more general solution using params (i.e. object arrays), the difference between that and adding a new status field on the surface doesn't feel like that big of a delta.

> I think the field we plan to add in status requires the Task to be trusted task

Does it? It doesn't seem that much different than the parameter based approach - it just gives us a more controlled/structured location to put provenance data in Run objects.

ywluogg replied:
Discussed offline with @wlynch and updated a note at the top of Proposal section!

wlynch commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,542 @@
+
+### DigestSets
+
+The in-toto provenance supports digest in terms of [DigestSets](https://github.com/in-toto/attestation/blob/main/spec/field_types.md#field-type-definitions) for an artifact, which means users can provide different digests for different digest SHAs. This would be hard to support in Chains, as Results doesn't support nested objects yet. In the proposal, only one pair of SHA and digest can be provid ...
```
```suggestion
The in-toto provenance supports digest in terms of [DigestSets](https://github.com/in-toto/attestation/blob/main/spec/field_types.md#field-type-definitions) for an artifact, which means users can provide different digests for different digest SHAs. Since Pipeline Results cannot support nested objects yet, Chains is unable to support them at the moment. In the proposal, only one pair of SHA and digest can be provided. `DigestSets` can be supported when nested objects are supported in `Results`.
```

Rephrased - it's not that this is hard for Chains to support, it's that we're limited by what Pipelines supports.

ywluogg commented:
> I can live with it, though I think we should change the arrays - I'm not sure if there's an easy way to have the Task provide these without hardcoding them. I don't think there's a way to derive the values from the [replacement context](https://tekton.dev/docs/pipelines/variables/#variables-available-in-a-task).

I just updated the schemas and removed the arrays

ywluogg commented:
/retest

wlynch commented on teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md:
```diff
@@ -0,0 +1,494 @@
+
+## Proposal
+
+**_Please note that the below proposal will be acted as a middle ground proposal, as it on one side definitely provides more organized structure for users to provide provenance data in Tekton TaskRuns and PipelineRuns, but on the other side that it still has a bunch of trade offs that can be resolved by another proposal mentioned in Alternatives section._**
```
It might help to refocus this TEP on the desired end state, and move the middle ground impl to `Implementation Plan` and/or `Risks and Mitigations`.

We should also look out to fill out the sections for `Drawbacks` - though I think this should focus on the end goal rather than in-between steps.

ywluogg commented:
/reopen

ywluogg commented:
/assign @wlynch
